File: sorting/problem49/problem49.cpp
#include <array>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

struct KeyHash {
    size_t operator()(const array<int, 26>& a) const noexcept {
        size_t h = 0;
        for (int v : a) {
            h = h * 1315423911u + static_cast<unsigned>(v + 1);
        }
        return h;
    }
};

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<array<int, 26>, vector<string>, KeyHash> mp;
        for (const string& s : strs) {
            array<int, 26> cnt{};
            for (char ch : s) {
                cnt[ch - 'a']++;
            }
            mp[cnt].push_back(s);
        }
        vector<vector<string>> res;
        res.reserve(mp.size());
        for (auto& kv : mp) {
            res.push_back(move(kv.second));
        }
        return res;
    }
};

File: sorting/problem49/problem49.py
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # Use 26-letter frequency as a hashable signature for anagrams
        groups: dict[tuple[int, ...], list[str]] = {}
        for s in strs:
            cnt = [0] * 26
            for ch in s:
                cnt[ord(ch) - ord('a')] += 1
            key = tuple(cnt)
            if key not in groups:
                groups[key] = []
            groups[key].append(s)
        return list(groups.values())
